src/rules/restrict-dependency-ranges.ts

```diff
@@ -0,0 +1,293 @@
+import type { AST as JsonAST } from "jsonc-eslint-parser";
+
+import * as ESTree from "estree";
+import semver from "semver";
+
+import { createRule } from "../createRule.js";
+import { isJSONStringLiteral } from "../utils/predicates.js";
+
+const DEPENDENCY_TYPES = [
+	"dependencies",
+	"devDependencies",
+	"optionalDependencies",
+	"peerDependencies",
+];
+
+interface Option {
+	forDependencyTypes?: string[];
+	forPackages?: string[];
+	forVersions?: string;
+	rangeType: RangeType | RangeType[];
+}
+
+type Options = [Option | Option[] | undefined];
+
+const RANGE_TYPES = ["caret", "pin", "tilde"];
+type RangeType = "caret" | "pin" | "tilde";
+
+const schemaOptions = {
+	additionalProperties: false,
+	properties: {
+		forDependencyTypes: {
+			items: {
+				enum: DEPENDENCY_TYPES,
+			},
+			type: "array",
+		},
+		forPackages: {
+			items: {
+				type: "string",
+			},
+			type: "array",
+		},
+		forVersions: {
+			type: "string",
+		},
+		rangeType: {
+			oneOf: [
+				{
+					enum: RANGE_TYPES,
+				},
+				{
+					items: {
+						enum: RANGE_TYPES,
+					},
+					type: "array",
+				},
+			],
+		},
+	},
+	required: ["rangeType"],
+	type: "object",
+};
+
+const SYMBOLS = {
+	caret: "^",
+	pin: "",
+	tilde: "~",
+} satisfies Record<RangeType, string>;
+
+/**
+ * Given the original version, update it to use the correct range type.
+ */
+const changeVersionRange = (version: string, rangeType: RangeType): string => {
+	// We need to handle workspace versions with only the range indicator,
+	// slightly differently
+	if (/^workspace:[~^*]$/.test(version)) {
+		switch (rangeType) {
+			case "caret":
+				return "workspace:^";
+			case "pin":
+				return "workspace:*";
+			case "tilde":
+			default:
+				return "workspace:~";
+		}
+	}
+
+	return version.replace(
+		/^(workspace:)?(\^|~|<=?|>=?)?/,
+		`$1${SYMBOLS[rangeType]}`,
+	);
+};
+
+/**
+ * Check if the version is in a form that this rule supports.
+ */
+const isVersionSupported = (version: string): boolean => {
+	if (version === "*" || /^workspace:[*^~]$/.test(version)) {
+		return true;
+	}
+
+	const rawVersion = version.replace(/^workspace:/, "");
+	return !!semver.valid(rawVersion) || !!semver.validRange(rawVersion);
```

Comment:
> `!!semver.valid(rawVersion) ||`

[Testing] No tests fail if this check is removed. Missing unit test(s)?

Reply:
Removed.  `semver.validRange` catches this.